File: ios/Sources/ACCore/ParseResult.swift
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

import Foundation

/// Warning generated during card parsing (non-fatal issues like unknown element types)
public struct ParseWarning: Sendable, Equatable {
    /// The type of warning
    public let code: Code
    /// Human-readable description
    public let message: String
    /// The JSON path where the warning occurred, if available
    public let path: String?

    public enum Code: String, Sendable {
        case unknownElementType
        case unknownActionType
        case missingInputId
        case fallbackUsed
    }

    public init(code: Code, message: String, path: String? = nil) {
        self.code = code
        self.message = message
        self.path = path
    }
}

/// Error types for card parsing failures
public enum ParseError: Error, Sendable, LocalizedError {
    case invalidJSON(String)
    case decodingFailed(String)
    case timeout
    case empty

    public var errorDescription: String? {
        switch self {
        case .invalidJSON(let message):
            return "Invalid JSON: \(message)"
        case .decodingFailed(let message):
            return "Decoding failed: \(message)"
        case .timeout:
            return "Parse operation timed out"
        case .empty:
            return "Empty JSON string"
        }
    }
}

/// Result of parsing an Adaptive Card JSON string.
/// Contains the parsed card (if successful), any warnings, and any error.
public struct ParseResult: Sendable {
    /// The parsed card, or nil if parsing failed
    public let card: AdaptiveCard?
    /// Non-fatal warnings encountered during parsing
    public let warnings: [ParseWarning]
    /// The error if parsing failed, nil on success
    public let error: ParseError?
    /// Time taken to parse in milliseconds (0 if served from cache)
    public let parseTimeMs: Double
    /// Whether the result was served from cache
    public let cacheHit: Bool

    /// Whether parsing succeeded (card is non-nil)
    public var isValid: Bool { card != nil }

    public init(
        card: AdaptiveCard? = nil,
        warnings: [ParseWarning] = [],
        error: ParseError? = nil,
        parseTimeMs: Double = 0,
        cacheHit: Bool = false
    ) {
        self.card = card
        self.warnings = warnings
        self.error = error
        self.parseTimeMs = parseTimeMs
        self.cacheHit = cacheHit
    }
}

File: android/ac-core/src/main/kotlin/com/microsoft/adaptivecards/core/ParseResult.kt
// Copyright (c) Microsoft Corporation. All rights reserved.
// Author: Vikrant Singh (github.com/VikrantSingh01)
// Licensed under the MIT License.

package com.microsoft.adaptivecards.core

import com.microsoft.adaptivecards.core.models.AdaptiveCard

/**
 * Warning generated during card parsing (non-fatal issues like unknown element types)
 */
data class ParseWarning(
    /** The type of warning */
    val code: Code,
    /** Human-readable description */
    val message: String,
    /** The JSON path where the warning occurred, if available */
    val path: String? = null
) {
    enum class Code {
        UNKNOWN_ELEMENT_TYPE,
        UNKNOWN_ACTION_TYPE,
        MISSING_INPUT_ID,
        FALLBACK_USED
    }
}

/**
 * Error types for card parsing failures
 */
sealed class ParseError(override val message: String) : Exception(message) {
    class InvalidJSON(msg: String) : ParseError("Invalid JSON: $msg")
    class DecodingFailed(msg: String) : ParseError("Decoding failed: $msg")
    class Timeout : ParseError("Parse operation timed out")
    class Empty : ParseError("Empty JSON string")
}

/**
 * Result of parsing an Adaptive Card JSON string.
 * Contains the parsed card (if successful), any warnings, and any error.
 */
data class ParseResult(
    /** The parsed card, or null if parsing failed */
    val card: AdaptiveCard? = null,
    /** Non-fatal warnings encountered during parsing */
    val warnings: List<ParseWarning> = emptyList(),
    /** The error if parsing failed, null on success */
    val error: ParseError? = null,
    /** Time taken to parse in milliseconds (0 if served from cache) */
    val parseTimeMs: Double = 0.0,
    /** Whether the result was served from cache */
    val cacheHit: Boolean = false
) {
    /** Whether parsing succeeded (card is non-null) */
    val isValid: Boolean get() = card != null
}
